### PR TITLE
dns error logging to diags

### PR DESCRIPTION
**Pull request by fabikd**
Logs using SiteThrottled* version is helpful for noisy logs, being used here instead of print statements to help with the possible retries when a dns code occurs

### CODE REVIEW

**Comment by ezelkow1** [comment]
[approve ci]

**Comment by bneradt** [comment]
@fabikd : can you please run `make -j8 format` and commit and push the changes? That should satisfy the clang-format CI job.

**Comment by bneradt** [comment]
[approve ci autest]

File: iocore/dns/DNS.cc [hunk 1]
```diff
@@ -1553,32 +1553,56 @@ dns_process(DNSHandler *handler, HostEnt *buf, int len)
...
   }
 
+  // Logs using SiteThrottled* version is helpful for noisy logs, being used here
+  // instead of print statements to help with the possible retries when a dns code occurs
   if (h->rcode != NOERROR || !h->ancount) {
     Debug("dns", "received rcode = %d", h->rcode);
     switch (h->rcode) {
     default:
-      Warning("Unknown DNS error %d for [%s]", h->rcode, e->qname);
+      SiteThrottledNote("Unknown: DNS error %d for [%s]", h->rcode, e->qname);
```
Comment (bneradt) [inline iocore/dns/DNS.cc:1562]: Maybe this should be kept a `SiteThrottledWarn` actually since this is unexpected.

**Review by bneradt — DISMISSED** [review]

File: iocore/dns/DNS.cc (hunk 1 above)
Comment (fabikd) [reply iocore/dns/DNS.cc]: I've changed the default to SiteThrolledWarning.

File: iocore/dns/DNS.cc [hunk 2]
```diff
@@ -1553,32 +1553,56 @@ dns_process(DNSHandler *handler, HostEnt *buf, int len)
...
       retry     = true;
       server_ok = false; // could be server problems
       goto Lerror;
-    case SERVFAIL: // recoverable error
+    case NOERROR: // NOERROR the if condition excludes this status code to avoid spamming the logs, good for testing
+      Debug("dns", "NOERROR: DNS code %d for [%s]: DNS Query completed successfully.", h->rcode, e->qname);
+      break;
+    case FORMERR: // FORMERR unrecoverable errors
+      SiteThrottledNote("FORMERR: DNS error %d for [%s]: DNS Query Format Error.", h->rcode, e->qname);
+      break;
```
Comment (bneradt) [inline iocore/dns/DNS.cc:1571]: `FORMERR` used to `goto Lerror`, now it breaks. I think since it is unrecoverable, we still want to `goto Lerror` right?

File: iocore/dns/DNS.cc [hunk 3]
```diff
@@ -1553,32 +1553,56 @@ dns_process(DNSHandler *handler, HostEnt *buf, int len)
...
     Debug("dns", "received rcode = %d", h->rcode);
     switch (h->rcode) {
     default:
-      Warning("Unknown DNS error %d for [%s]", h->rcode, e->qname);
+      SiteThrottledWarning("Unknown: DNS error %d for [%s]", h->rcode, e->qname);
       retry     = true;
       server_ok = false; // could be server problems
       goto Lerror;
-    case SERVFAIL: // recoverable error
+    case NOERROR: // NOERROR the if condition excludes this status code to avoid spamming the logs, good for testing
```
Comment (bneradt) [inline iocore/dns/DNS.cc:1566]: I suggest rewording the comment to something like:

```
// Included for completeness. The above condition ensures that NOERROR should not enter this block.

File: iocore/dns/DNS.cc [hunk 4]
```diff
@@ -1553,32 +1553,56 @@ dns_process(DNSHandler *handler, HostEnt *buf, int len)
...
+      Debug("dns", "NOERROR: DNS code %d for [%s]: DNS Query completed successfully.", h->rcode, e->qname);
+      break;
+    case FORMERR: // FORMERR unrecoverable errors
```
Comment (bneradt) [inline iocore/dns/DNS.cc:1569]: For each of these cases, repeating the value name (`FORMERR, `SERVFAIL`, etc.) in the comments is unnecessary duplication. Just remove them from the comments.

File: iocore/dns/DNS.cc [hunk 5]
```diff
@@ -1553,32 +1553,56 @@ dns_process(DNSHandler *handler, HostEnt *buf, int len)
...
       break;
-    case FORMERR: // unrecoverable errors
-    case REFUSED:
-    case NOTIMP:
-      Debug("dns", "DNS error %d for [%s]", h->rcode, e->qname);
+    case NXDOMAIN: // NXDOMAIN
+      SiteThrottledNote("NXDOMAIN: DNS error %d for [%s]: Domain name does not exist.", h->rcode, e->qname);
+      retry     = false;
+      tcp_retry = false;
+      break;
```
Comment (bneradt) [inline iocore/dns/DNS.cc:1580]: This also used to `goto Lerror` and now doesn't with this patch. Is that intentional?

File: iocore/dns/DNS.cc [hunk 6]
```diff
@@ -1553,32 +1553,56 @@ dns_process(DNSHandler *handler, HostEnt *buf, int len)
...
+    case YXRRSET: // YXRRSET
+      SiteThrottledNote("YXRRSET: DNS error %d for [%s]: RRSet that should not exist, does exist.", h->rcode, e->qname);
+      break;
+    case NXRRSET: // NXRRSET
+      SiteThrottledNote("NXRRSET: DNS error %d for [%s]: RRSet that should not exist, does not exist.", h->rcode, e->qname);
+      break;
+    case NOTAUTH: // NOTAUTH
+      SiteThrottledNote("NOTAUTH: DNS error %d for [%s]: Server not authorative for the zone.", h->rcode, e->qname);
+      break;
```
Comment (bneradt) [inline iocore/dns/DNS.cc:1601]: Each of these used to `goto Lerror` before this patch. Can you please indicate whether this change in behavior in this patch is intentional for these? Judging from the description, I expected this to be a logging-only change patch, not a DNS functional change.

**Review by bneradt — CHANGES_REQUESTED** [review]
Thank you for updating the logging for DNS. That can be helpful for debugging our DNS implementation when it is behaving in a way we don't expect.

The behavior for many of these cases changes with this patch. Can you please provide some comments in the PR for the rationale behind this?

My understanding is that the cases that simply break are "recoverable" and the ones that `goto Lerror` are not recoverable. I'm not familiar enough offhand to tell which case belongs to which. I'd like for @SolidWallOfCode to take a look at how these are now being categorized to verify that these changes are OK. He's more familiar with DNS.cc logic than I.

**Comment by fabikd** [comment]
Right, I think it's a great idea to have @SolidWallOfCode look at it. I used the code from the master branch, https://github.com/apache/trafficserver/blob/master/iocore/dns/DNS.cc#L1556
should I have used a different version? I’ve looked at the existing code in the master branch and these 3 labels did show `goto Lerror`: default, NOTIMP, NOTZONE, and my intent was to leave `goto Lerror` as is. I do not see `goto Lerror` on this version of the code https://github.com/apache/trafficserver/blob/master/iocore/dns/DNS.cc#L1556 for FORMERR, NXDOMAIN, YXDOMAIN, YXRRSET, NXRRSET, NOTAUTH and it was not my intention to remove `goto Lerror` on any label.

As for the breaks, usually a switch statement has breaks after each case to avoid passing through the code under the next label. If I am using the correct version of the code, was the intention of it to not break and pass through all the named labels and use the last `goto Lerror` as a catch all? 
https://github.com/apache/trafficserver/blob/master/iocore/dns/DNS.cc#L1581 
If so, I can add `goto Lerror` to FORMERR, NXDOMAIN, YXDOMAIN, YXRRSET, NXRRSET, NOTAUTH. Just let me know! Thank you!! (:

File: iocore/dns/DNS.cc (hunk 2 above)
Comment (fabikd) [reply iocore/dns/DNS.cc:1571]: please refer to my comment below.

File: iocore/dns/DNS.cc (hunk 3 above)
Comment (fabikd) [reply iocore/dns/DNS.cc:1566]: will do!

File: iocore/dns/DNS.cc (hunk 4 above)
Comment (fabikd) [reply iocore/dns/DNS.cc:1569]: will do!!

File: iocore/dns/DNS.cc (hunk 5 above)
Comment (fabikd) [reply iocore/dns/DNS.cc:1580]: please refer to my comment below.

File: iocore/dns/DNS.cc (hunk 6 above)
Comment (fabikd) [reply iocore/dns/DNS.cc:1601]: please refer to my comment below.

**Comment by bneradt** [comment]
>  I used the code from the master branch, https://github.com/apache/trafficserver/blob/master/iocore/dns/DNS.cc#L1556 should I have used a different version?

You're PR is correctly based on the `master` branch. Nothing to change there.

> I’ve looked at the existing code in the master branch and these 3 labels did show `goto Lerror`: default, NOTIMP, NOTZONE, and my intent was to leave `goto Lerror` as is. I do not see `goto Lerror` on this version of the code https://github.com/apache/trafficserver/blob/master/iocore/dns/DNS.cc#L1556 for FORMERR, NXDOMAIN, YXDOMAIN, YXRRSET, NXRRSET, NOTAUTH and it was not my intention to remove `goto Lerror` on any label.
> 
> As for the breaks, usually a switch statement has breaks after each case to avoid passing through the code under the next label. 

Developers must intentionally `break`, `return`, `goto`, or not based on the flow of code they desire. Without statements such as these, the runtime path will "fall through" to the following case statements.

> If I am using the correct version of the code, was the intention of it to not break and pass through all the named labels and use the last `goto Lerror` as a catch all? https://github.com/apache/trafficserver/blob/master/iocore/dns/DNS.cc#L1581 If so, I can add `goto Lerror` to FORMERR, NXDOMAIN, YXDOMAIN, YXRRSET, NXRRSET, NOTAUTH. Just let me know! Thank you!! (:

Correct. Unless `break` statements or the like are used, `case` statements fall through to next ones. Thus, in this block from master branch (before your change):

```
    case NOERROR:
    case NXDOMAIN:
    case 6:  // YXDOMAIN
    case 7:  // YXRRSET
    case 8:  // NOTAUTH
    case 9:  // NOTAUTH
    case 10: // NOTZONE
      Debug("dns", "DNS error %d for [%s]", h->rcode, e->qname);
      goto Lerror;
```

All of `NOERROR`, `NXDOMAIN`, etc., will fall through to that same `Debug` log line and `goto Lerror`. There's no need to repeat the `Debug` and `goto Lerror` statements for each case.

This can be seen with the following short example:
https://godbolt.org/z/cqodTTKvq

Thank you again for working on this.

**Comment by SolidWallOfCode** [comment]
It is also very standard (from the early days of C coding, which I remember personally) to omit `break` for cases that have completely identical code, as in this situation.

Also note that for C++17 there is now an standard keyword that should (must?) be used for fallthrough in `switch` statements. E.g.

```
switch (value) {
  case 0:
    // code ...
    [[fallthrough]];
  case 1:
    // code ...
```
See [here](https://en.cppreference.com/w/cpp/language/attributes/fallthrough). Note the style for `case 1:` and `case 2:` in the example.

**Comment by SolidWallOfCode** [comment]
That said, the errors are a bit odd, particularly this one
```
         - 502 https://127.0.0.1:``/ 2 0
         ?                       ^^
         
         + 502 https://127.0.0.1:61712/ 2 0
         ?                       ^^^^^
```
My understanding was the double back tick indicated "anything goes" for that token. So what is the failure, exactly? The cert checking logic is a bit dependent on DNS due to how TLS vs. TCP connection errors are handled. It might be that the changes in the query reply code handling is causing this problem.

**Comment by SolidWallOfCode** [comment]
This may be related to the problem:

On master -
```
    case NXDOMAIN:
    case 6:  // YXDOMAIN
    case 7:  // YXRRSET
    case 8:  // NOTAUTH
    case 9:  // NOTAUTH
    case 10: // NOTZONE
      Debug("dns", "DNS error %d for [%s]", h->rcode, e->qname);
      goto Lerror;
```

In the PR
```
    case NXDOMAIN:
      SiteThrottledNote("NXDOMAIN: DNS error %d for [%s]: Domain name does not exist.", h->rcode, e->qname);
      retry     = false;
      tcp_retry = false;
      goto Lerror;
      break;
```

Note the change in handling of `retry` and `tcp_retry`. This shouldn't happen for a PR that is only changing the logging.

**Comment by fabikd** [comment]
I returned the switch to its original order instead of chronological order, removed all the breaks, and retry in order to allow for the original intention of allowing the fallthrough. In theory this will print the statement that it is selected as well as all the subsequent statements, that comes after it. example: 
 
`switch (1){
    case 0:
std::cout << 0;
    case 1:
std::cout << 1;
    case 2:
std::cout << 2;
    case 3:
std::cout << 3;`
Output: `123`
 
When researching the logs, it would be confusing to see `123` and have to know that the log that applies is the 1st one. Do we still want to print to log in this case?

Also, I believe I can add `[[fallthrough]];` after the code in each label and it might build the failed tests on Rocky, Fedora, Debian, and AuTest.  Because the Diags.h file is flagging all the fall through and failing them.

**Comment by maskit** [comment]
I'm just a passerby, so I'm sorry if I'm missing some background or previous discussion, but it seems like to me that all the log messages have the same format and you could avoid the repeat. Since DNS RCODE is basically an integer from 0 to 23, I'd simply replace the current `Debug` lines with:

```cpp
SiteThrottledNote("%s: DNS error %d for [%s]: %s", RCODE_NAME[h->rcode], h->rcode, e->qname, RCODE_DESCRIPTION[h->rcode]);
```

I guess using some macros, `constexpr`, or `Lexicon` from libswoc would make this a bit more efficient.

https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#dns-parameters-6

**Comment by fabikd** [comment]
> I'm just a passerby, so I'm sorry if I'm missing some background or previous discussion, but it seems like to me that all the log messages have the same format and you could avoid the repeat. Since DNS RCODE is basically an integer from 0 to 23, I'd simply replace the current `Debug` lines with:
> 
> ```c++
> SiteThrottledNote("%s: DNS error %d for [%s]: %s", RCODE_NAME[h->rcode], h->rcode, e->qname, RCODE_DESCRIPTION[h->rcode]);
> ```
> 
> I guess using some macros, `constexpr`, or `Lexicon` from libswoc would make this a bit more efficient.
> 
> https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#dns-parameters-6

Thank you for your feedback!! I will give it a try! (:

**Comment by bryancall** [comment]
[approve ci autest]

File: iocore/dns/DNS.cc [hunk 7]
```diff
@@ -1553,31 +1573,36 @@ dns_process(DNSHandler *handler, HostEnt *buf, int len)
...
     switch (h->rcode) {
     default:
-      Warning("Unknown DNS error %d for [%s]", h->rcode, e->qname);
+      SiteThrottledWarning("UNKNOWN: DNS error %d for [%s]", h->rcode, e->qname);
       retry     = true;
       server_ok = false; // could be server problems
       goto Lerror;
+    case NOERROR: // Included for completeness. The above condition ensures that NOERROR should not enter this block.
+      Debug("dns", "%s: DNS error %d for [%s]: %s", RCODE_NAME[h->rcode], h->rcode, e->qname, RCODE_DESCRIPTION[h->rcode]);
+      break; // need to break here or this fails test: proxy_serve_stale_dns_fail even though NOERROR is not used in this block.
```
Comment (bneradt) [inline iocore/dns/DNS.cc:1588]: I see. The NOERROR can enter here if `!h->ancount` is true. So this comment, which I think I suggested to you, actually doesn't make sense.

I suggest removing the two comments you've added on these lines but leave the case handling NOERROR. I assume changing `NOERROR` to a Debug is a good change since we don't need to warn about it.

File: iocore/dns/DNS.cc [hunk 8]
```diff
@@ -32,6 +32,8 @@
 #define SRV_SERVER (RRFIXEDSZ + 6)
 #define SRV_FIXEDSZ (RRFIXEDSZ + 6)
 
+#define RCODE_ARRAY_SIZE 11
+
```
Comment (bneradt) [inline iocore/dns/DNS.cc:36]: We don't need `RCODE_ARRAY_SIZE`. If we initialize an array without an explicit size, the compiler figures out the size for us. Thus:

```
  const char *RCODE_NAME[] = {
     "NOERROR", "FORMERR", "SERVFAIL", "NXDOMAIN", "NOTIMP", "REFUSED", "YXDOMAIN", "YXRRSET", "NXRRSET", "NOTAUTH", "NOTZONE",
   };
```

**Review by bneradt — CHANGES_REQUESTED** [review]
Looks good. Just a couple suggestions.

File: iocore/dns/DNS.cc (hunk 8 above)
Comment (fabikd) [reply iocore/dns/DNS.cc:36]: Sounds good! (:

**Review by bneradt — APPROVED** [review]
Looks good.

**Comment by SolidWallOfCode** [comment]
Seems fine. A couple of notes, although not worth delaying the merge.

For the names, we might consider using [Lexicon](http://docs.solidwallofcode.com/libswoc/code/Lexicon.en.html), although the utility is less because it's a one way conversion. I might be tempted to have an array of string pairs, to group the tag and description, but then the use site is more complex. This is pretty deep in the style weeds so take them as potential ideas for future work.

**Review by SolidWallOfCode — APPROVED** [review]

**Review by maskit — APPROVED** [review]
